jaxrs-api/src/main/java/javax/ws/rs/JAXRS.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2018 Markus KARG. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package javax.ws.rs;
+
+import java.util.function.Function;
+
+import javax.ws.rs.core.Application;
+import javax.ws.rs.ext.RuntimeDelegate;
+
+/**
+ * Bootstrap class that is used to startup a JAX-RS application in Java SE
+ * environments.
+ *
+ * The {@code JAXRS} class is available in a Jakarta EE container environment as
+ * well; however, support for the Java SE bootstrapping APIs is <em>not
+ * required</em> in container environments.
+ *
+ * @author Markus KARG (markus@headcrashing.eu)
+ *
+ * @since 2.2
+ */
+public interface JAXRS {
+
+    /**
+     * Invoked in Java SE environments to start the provided application at the
+     * specified root URL.
+     *
+     * This method will not return until the JAX-RS application is terminated.
```

Comment:
How is a JAX-RS application terminated?  Would you want to add a new `stop()` or `close()` method to the `Application` class?  Alternatively, maybe the `JAXRS.start` method could return immediately after starting the application asynchronously, and then you could add a `JAXRS.stop` method to terminate the application?

Reply:
My PR only proposed to wrap existing vendor-specific technology under a common API hood. AFAIK not all JAX-RS runtimes do support explicitly stopping an application. So your proposal would be to add a completely new feature. If all vendors support that, I am fine with it. But until then, I cannot unilaterally set up a PR for that.

So question to @jansupol, @asoldano, @deki: Do you like to add support for explicitly stopping a JAX-RS application as proposed by @andymc12? I would be more than happy to add this to my PR in that case.

Reply:
Hi @mkarg - I'm mainly having trouble understanding how the `start()` method works - or more specifically how the method ever returns.  And possibly my brain is malfunctioning because I've been so deep in the Java EE (server-side) mode for too long that I can't process Java SE patterns.  ;-)  

Here are a couple scenarios where I think a blocking `start` method with no spec-defined way to stop the app could be troublesome:
1) User wants an app to read from a database.  In their `main` method, they set up some database connections then call `start` to load up the JAX-RS front-end.  At some point, they will want to stop the process (perhaps to deploy a new version of the app, or do DB maintenance, etc.).  They could just kill the JVM process, but that won't gracefully disconnect from the DB - not a huge problem, but still messy.

2) User wants to test their app before deploying it to a Java EE server.  They set up mock resources (like WireMock or MockServer, etc.) that run in separate processes.  Once their testing is complete, the user would need to kill not only their Java SE JAX-RS process, but any other processes that it may have started - instead of leveraging the convenient APIs for stopping those processes.

I'm not sure if CXF has a way to stop an application (outside of the runtime it is running in - Tomcat, OpenLiberty, etc.) so it might not have a mechanism to return from this method as it is today.  That's not necessarily a problem - but then we should probably update the spec to indicate that implementations should provide a mechanism for stopping apps.  

Hope this helps...

Reply:
@andymc12 I share your thoughts, but again, I am only trying to wrap the current behaviour of the existing products as-is. The start method will only be implemented IF the existing products CAN be started on Java SE (this is AFAIK not mandatory as of JAX-RS 2.1+). IF an existing product hasn't a way to be stopped BUT supports Java SE, then effectively effects 1 and 2 will happen actually, independent of my proposed API or not -- again, I am not specifying a new feature or behaviour, I only wrap EXISTING products. So again my question to all vendors: Can your existing products be stopped explicitly, yes or no? If yes, I can add a stop mechanism in this API. If no, are you willing to change this?

Reply:
In think in most cases ShutdownHook is used in this case and CXF is shutdown by closing the Spring context. Should be also possible to implement it for non Spring scenarios, org.apache.cxf.endpoint.Server has stop and destroy methods. Comparing to JAX-WS javax.xml.ws.Endpoint has also a stop method, so I'm open to have something similar in JAX-RS.